### ISSUE TITLE
int128.h incompatible with MSVC

### ISSUE THREAD

**Issue by Photosounder** [WasmEdge/WasmEdge#2419]
I'm trying to use 0.12.0-alpha.2's `WasmEdge-0.12.0-windows.msi` on MSVC. I suspect that no one has tried to use WasmEdge on Visual Studio in a long time or they would have had the same problem:

```
1>------ Build started: Project: wasm_host, Configuration: Debug x64 ------
1>wasm_host.c
1>C:\Program Files\WasmEdge\include\wasmedge\int128.h(19,18): error C4235: nonstandard extension used: '__int128' keyword not supported on this architecture
1>C:\Program Files\WasmEdge\include\wasmedge\int128.h(20,9): error C4235: nonstandard extension used: '__int128' keyword not supported on this architecture
```

I'm quite puzzled by the `#if` statement in int128.h:
```c
#if defined(__x86_64__) || defined(__aarch64__)
typedef unsigned __int128 uint128_t;
typedef __int128 int128_t;
#else
```

The latest version of this file uses `#if defined(__x86_64__) || defined(__aarch64__) || (defined(__riscv) && __riscv_xlen == 64)` which is no better. Maybe just using `#if defined(__SIZEOF_INT128__)` would be preferable.

**Comment by Photosounder** [comment]
I investigated the problem further and it's a pretty tricky one, changing `include/api/wasmedge/int128.h` alone doesn't cut it, there's also `include/common/int128.h`. The problem with the latter is that somehow it has no problem with `__int128`, so then C++ and C code use different definitions of `int128_t` which somehow aren't even the same size which causes other bugs I reported (#2421 and #2422).

If at line 18 of `include/common/int128.h` we use `#if 0` then WasmEdge doesn't even compile anymore because:
```
C:\msys\home\lib\WasmEdge\include\common/int128.h(28,10): fatal error: 'boost/predef/other/endian.h' file not found
#include <boost/predef/other/endian.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
and even if we comment that out and remove the endianness check the rest doesn't compile anyway, there are tons of errors which makes me think this code hasn't been used by anyone in a very long time (seems like the lack of testing on certain platforms is a recurring problem with this project).

So somehow the `int128_t` of `include/api/wasmedge/int128.h` has to be reconciled with the one from `include/common/int128.h`

**Comment by abhinandanudupa** [comment]
The `...smedge\int128.h(19,18): error C4235: nonstandard extension...` in the error suggests to be that you are not using the right compiler. WasmEdge uses clang for all libraries and dependencies on Windows (atleast as far as I know). Please refer to this - https://wasmedge.org/book/en/contribute/build_from_src/windows.html for instructions on how to build on Windows.

If you want to use VS to build WasmEdge you will have to edit the configuration as by default VS forces MSVC on projects. You can change the environment by clicking on the arrow below the configuration and following appropriate options (which I have forgotten). Note that you will also have to install clang build tools in VS before you have options for chaning the environment.

`x86_clang`, `x64_64_clang` are some options to help you find the right menu.

**Comment by Photosounder** [comment]
I don't use Clang to compile my host and I'm not going to, a library doesn't dictate what compiler you use. It's pretty ridiculous that this library even needs Clang to build btw, but not surprising when code like the one highlighted in the OP doesn't even take MSVC into account, like no one ever tried that before.

**Comment by hydai** [comment]
Hi @Photosounder 
Thanks for the information. We don't try MSVC before, so the header `wasmedge.h` may have something incompatible with MSVC. Unfortunately, it does happen.
And that's right, the `__int128` is an extension of the integer scalar type, which is only provided by GCC and Clang. We will change the check condition based on what you recommended. Thanks!

**Comment by hydai** [comment]
This should be fixed in #2751
